x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/ClusterDeprecationChecks.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.deprecation;
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.indices.ShardLimitValidator;
+import org.elasticsearch.xpack.core.deprecation.DeprecationIssue;
+
+public class ClusterDeprecationChecks {
+    static DeprecationIssue checkShards(ClusterState clusterState) {
+        // Make sure we have room to add a small non-frozen index if needed
+        if (ShardLimitValidator.canAddShardsToCluster(5, 1, clusterState, false)) {
+            return null;
+        } else {
+            return new DeprecationIssue(
+                DeprecationIssue.Level.WARNING,
+                "The cluster has too many shards to be able to upgrade",
+                "https://ela.st/es-deprecation-7-shard-limit",
+                "Delete indices to clear up space",
```

Comment:
Is this message from another deprecation error? Otherwise maybe we can be more specific and say something like "there is not enough room for N more shards, increase the `cluster.max_shards_per_node` or `cluster.max_shards_per_node.frozen` settings or remove indices to clear up resources."

(using "space" is hard here, because it makes it sound like it's disk-related, when really it's just our arbitrary limits)

Reply:
It might be used in another error message, but in this case I think it was just me trying to walk the line between being too technical and not technical enough to be useful. I'll reword it.

Reply:
I intentionally didn't include `cluster.max_shards_per_node.frozen` in the reworded version because we're not checking that setting (the assumption is that the new small index would not be a frozen index).